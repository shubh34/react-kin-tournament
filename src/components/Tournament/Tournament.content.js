const content = {
	gainNexAvaliablePrizeText: 'Gain :next-avail-prize points and win the next prize!',
};

export default content;
