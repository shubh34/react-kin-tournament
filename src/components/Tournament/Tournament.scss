@import '../../colors';

.c-tournament {
    background-color: $darkGray;
    padding: 32px 40px 24px 40px;
}