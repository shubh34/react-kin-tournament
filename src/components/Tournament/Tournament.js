import './Tournament.scss';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchTournamentLeaders, fetchTournamentPrizes } from '../../states/tournament/actions';
import {
	getPrizes, getTopLeaders, isUserInTopPrizeRange, getUserNextAvailabelPrize, isUserPlayingTournament,
} from '../../states/tournament/selectors';
import { API_CACHE_TIMER } from '../../configs/config';
import Tabs from '../../sharedComponent/Tabs';
import LeaderBoard from '../Leaderboard/LeaderBoard';
import Prizes from '../Prizes/Prizes';
import { getUserId } from '../../states/userDetails/selectors';
import content from './Tournament.content';

export const tourmamentTabs = {
	LEADERBOARD: 'Leaderboard',
	PRIZES: 'Prizes',
};
const mapDispatch = dispatch => ({
	fetchLeaders: () => dispatch(fetchTournamentLeaders()),
	fetchPrizes: () => dispatch(fetchTournamentPrizes()),
});
const mapState = state => ({
	userId: getUserId(state),
	prizes: getPrizes(state),
	leaders: getTopLeaders(state),
	isUserPlayingTournament: isUserPlayingTournament(state),
	isUserInTopPrizeRange: isUserInTopPrizeRange(state),
	userNextAvailabelPrize: getUserNextAvailabelPrize(state),
});

// Question : should we clear setInterval on unmount

export class Tournament extends Component {
	constructor(props) {
		super(props);
		this.state = {
			activeTab: tourmamentTabs.LEADERBOARD,
		};
		this.fetchTournamentDetails = this.fetchTournamentDetails.bind(this);
		this.onTablinkChange = this.onTablinkChange.bind(this);
	}


	componentDidMount() {
		this.fetchTournamentDetails();
		setInterval(this.fetchTournamentDetails, API_CACHE_TIMER);
	}

	onTablinkChange(event) {
		this.setState({
			activeTab: event.target.name,
		});
	}

	fetchTournamentDetails() {
		const { fetchLeaders, fetchPrizes } = this.props;
		fetchLeaders();
		fetchPrizes();
	}

	render() {
		const { activeTab } = this.state;
		const {
			leaders, userNextAvailabelPrize, isUserInTopPrizeRange, userId, prizes, isUserPlayingTournament,
		} = this.props;
		const isLeaderboardActive = activeTab === tourmamentTabs.LEADERBOARD;
		const userNextAvailabelPrizeText = content.gainNexAvaliablePrizeText.replace(':next-avail-prize', userNextAvailabelPrize);
		const showNextAvailablePrizeText = isUserPlayingTournament && !isUserInTopPrizeRange;
		return (
			<div className="c-tournament">
				<Tabs>
					<Tabs.TabsLink name={tourmamentTabs.LEADERBOARD} onClick={this.onTablinkChange} isActive={isLeaderboardActive} />
					<Tabs.TabsLink name={tourmamentTabs.PRIZES} onClick={this.onTablinkChange} isActive={!isLeaderboardActive} />
					<Tabs.TabsContent>
						{isLeaderboardActive && <LeaderBoard leaders={leaders} userId={userId} />}
						{!isLeaderboardActive && <Prizes prizes={prizes} />}
						{showNextAvailablePrizeText && <p className="c-tournament-information">{userNextAvailabelPrizeText}</p>}
					</Tabs.TabsContent>
				</Tabs>
			</div>
		);
	}
}
Tournament.propTypes = {
	fetchLeaders: PropTypes.func.isRequired,
	fetchPrizes: PropTypes.func.isRequired,
	leaders: PropTypes.arrayOf(PropTypes.shape({
		position: PropTypes.number.isRequired,
		score: PropTypes.number.isRequired,
		prize: PropTypes.string.isRequired,
	})),
	prizes: PropTypes.arrayOf(PropTypes.shape({
		prize: PropTypes.string.isRequired,
		range: PropTypes.string.isRequired,
	})),
	userNextAvailabelPrize: PropTypes.number.isRequired,
	isUserInTopPrizeRange: PropTypes.bool.isRequired,
	isUserPlayingTournament: PropTypes.bool.isRequired,
	userId: PropTypes.number.isRequired,
};

export default connect(mapState, mapDispatch)(Tournament);
