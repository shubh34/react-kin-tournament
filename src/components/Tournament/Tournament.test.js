import configureStore from 'redux-mock-store';
import React from 'react';
import { shallow, mount } from 'enzyme';

import Tournament from './Tournament';
import { getMockState } from '../../../tests/MockState';

const mockStore = configureStore();

const setUp = (state = 'DEFAULT', setUpProps = {}, container = shallow) => {
	const props = {
		...setUpProps,
	};
	const store = mockStore(getMockState(state));
	store.dispatch = jest.fn();
	const wrapper = container(<Tournament {...props} store={store} />);
	return { wrapper, props, store };
};

describe('Tournament', () => {
	it('renders without crashing', () => {
		const { wrapper } = setUp();
		expect(wrapper.dive().dive()).toMatchSnapshot();
	});
	it('component did mount should call apis', () => {
		const { store, wrapper } = setUp(undefined, {}, mount);
		expect(store.dispatch).toBeCalled();
		expect(wrapper).toMatchSnapshot();
	});
});
