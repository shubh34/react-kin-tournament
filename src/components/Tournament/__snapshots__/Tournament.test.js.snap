// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tournament renders without crashing 1`] = `
<div
  className="c-tournament"
>
  <Tabs>
    <TabsLink
      isActive={true}
      name="Leaderboard"
      onClick={[Function]}
    />
    <TabsLink
      isActive={false}
      name="Prizes"
      onClick={[Function]}
    />
    <TabsContent>
      <LeaderBoard
        leaders={
          Array [
            Object {
              "playerId": 3,
              "position": 1,
              "prize": "10836.00",
              "score": 87,
            },
            Object {
              "playerId": 18,
              "position": 2,
              "prize": "10836.00",
              "score": 86,
            },
            Object {
              "playerId": 2,
              "position": 3,
              "prize": "5483.00",
              "score": 75,
            },
            Object {
              "playerId": 1,
              "position": 7,
              "prize": "0.00",
              "score": 28,
            },
          ]
        }
        userId={1}
      />
      <p
        className="c-tournament-information"
      >
        Gain 44 points and win the next prize!
      </p>
    </TabsContent>
  </Tabs>
</div>
`;

exports[`Tournament should active prizes when prize tab is clicked 1`] = `
<div
  className="c-tournament"
>
  <Tabs>
    <TabsLink
      isActive={false}
      name="Leaderboard"
      onClick={[Function]}
    />
    <TabsLink
      isActive={true}
      name="Prizes"
      onClick={[Function]}
    />
    <TabsContent>
      <Prizes
        prizes={
          Array [
            Object {
              "fromPosition": 1,
              "prize": 10836,
              "range": "1",
              "toPosition": 2,
            },
            Object {
              "fromPosition": 3,
              "prize": 5483,
              "range": "2-3",
              "toPosition": 3,
            },
            Object {
              "fromPosition": 4,
              "prize": 1079,
              "range": "4-5",
              "toPosition": 5,
            },
          ]
        }
      />
      <p
        className="c-tournament-information"
      >
        Gain 44 points and win the next prize!
      </p>
    </TabsContent>
  </Tabs>
</div>
`;
